dynamicio/core.py

```diff
@@ -100,6 +101,20 @@ async def async_read(self):
         loop = asyncio.get_running_loop()
         return await loop.run_in_executor(pool, self.read)
 
+    @cached_property
+    def df(self) -> pd.DataFrame:
+        """Uses read() and caches it. Useful for lazy-loading.
+
+        Returns:
+            A pandas dataframe or an iterable.
+        """
+        return self.read()
+
+    def clear_cache(self) -> None:
+        """Resets self.df cached_property by deleting it."""
+        if "df" in self.__dict__:
+            del self.df
+
```

Comment:
There are so many things to discuss here. 

Firstly, I am not sure that this feature is necessary at all.  I do appreciate, however, that the existence of it, will promote a practice. Let me explain. 

Suppose we have the following code: 

```python
import os

from demo.src import RESOURCES, ENVIRONMENT, environment
from dynamicio import UnifiedIO
from dynamicio.config import IOConfig
from dynamicio.core import SCHEMA_FROM_FILE


class InputIO(UnifiedIO):
    """UnifiedIO subclass for V6 data."""

    schema = SCHEMA_FROM_FILE

input_config = IOConfig(
    path_to_source_yaml=(os.path.join(RESOURCES, "definitions/input.yaml")),
    env_identifier=ENVIRONMENT,
    dynamic_vars=environment,
)

bar_df = InputIO(source_config=input_config.get(source_key="BAR"), apply_schema_validations=True, log_schema_metrics=True).read()

# do some other stuff
# ...now process bar_df
# write output
# finish
```
This is a standardised way of working. 

Arguably, you make a point here saying that "I don't want to load `bar_df` at the very start of my module; this will occupy memory unnecessarily. I want to load it only when I need it". 

Well, that's fine as long as you modify your code as follows: 
 ```python
"""Add module docstring...."""
import os

from demo.src import RESOURCES, ENVIRONMENT, environment
from dynamicio import UnifiedIO
from dynamicio.config import IOConfig
from dynamicio.core import SCHEMA_FROM_FILE


class InputIO(UnifiedIO):
    """UnifiedIO subclass for V6 data."""

    schema = SCHEMA_FROM_FILE

input_config = IOConfig(
    path_to_source_yaml=(os.path.join(RESOURCES, "definitions/input.yaml")),
    env_identifier=ENVIRONMENT,
    dynamic_vars=environment,
)

bar_io = InputIO(source_config=input_config.get(source_key="BAR"), apply_schema_validations=True, log_schema_metrics=True)

# do some other stuff
# ...now process bar_df
bar_df = bar_io.read()
# write output
# finish
```
This way of working has recently been demonstrated by @tylerferguson.  Also, once read in, `bar_df` can then be treated equivalently to your cached property (`df`). So you can either modify it in place, as you illustrate in your tests, or avoid doing so in case you want to use the original downstream.  

So, I am asking myself, "what is the value of having df as a `cached_property`"? The way I see it is that the above code would be rewritten as follows: 
```python
"""Add module docstring...."""
import os

from demo.src import RESOURCES, ENVIRONMENT, environment
from dynamicio import UnifiedIO
from dynamicio.config import IOConfig
from dynamicio.core import SCHEMA_FROM_FILE


class InputIO(UnifiedIO):
    """UnifiedIO subclass for V6 data."""

    schema = SCHEMA_FROM_FILE

input_config = IOConfig(
    path_to_source_yaml=(os.path.join(RESOURCES, "definitions/input.yaml")),
    env_identifier=ENVIRONMENT,
    dynamic_vars=environment,
)

bar = InputIO(source_config=input_config.get(source_key="BAR"), apply_schema_validations=True, log_schema_metrics=True)

# do some other stuff
# ...now process bar_df
df = do_stuff_to(bar.df)
# write output
# finish
```

I would argue the two latter approaches are equivalent in style. The one you are pitching here is indeed more elegant but it worries me too. This is because we are suddenly offering developers an alternative way of loading data. The premise is righteous, but then again `read()` remains part of the public interface (note here that you are not accounting for `async_read()`, which needs to be addressed if we eventually decide to merge this in) and thus, developers will continue to use it. This flexibility is unnecessary in my opinion. if anything, I would be more than happy to remove `read()` altogether and only allow the use of `.df` as a cached property. Personally, I feel like `dynamicio` should be as rigid in how it is used as possible, to promote consistency. 

Let me emphasize my point clearly here. I don't disagree with what you recommend. I disagree with introducing it as an alternative way of loading data. Let's instead agree on one way of loading data and provide alternatives if imperative. 

So, I think this requires a broader discussion and that's why I am involving @tylerferguson and @VRGhost in the discussion.  

Another point worth discussing is how `df` is now a property. Again, nothing wrong with this, but it does change the philosophy of how `dynamicio` is used. See, `InputIO`, which is defined as follows: 

```python

class InputIO(UnifiedIO):
    """UnifiedIO subclass for V6 data."""

    schema = SCHEMA_FROM_FILE
```

has a very specific responsibility: "To return or write a dataframe". You are now changing it to an instance that will be hosting one. So, maybe inadvertently you are engaging in a discussion about what should be the responsibility of `UnifiedIO` subclasses. This is neither right or wrong, but it is fundamentally different to how these instances are used. As a matter of fact, assuming we accept this PR, then we will be able to do things like: 
```python

class InputIO(UnifiedIO):
    """UnifiedIO subclass for V6 data."""

    schema = SCHEMA_FROM_FILE
    
    def modify_df(self):
        self.df = self.df.reset_index(drop=True)     

    def some_other_modification(self):
        ...
```

You may be thinking, what's wrong with this? Again, nothing. I actually do like it more. But again it disagrees with the framework's current philosophy--`dynamicio` just does I/O; it does not host data. Furthermore, things get a bit trickier too if you go in this direction. This is because currently, a single `InputIO(UnifiedIO)` instance can be used to load all possible datasets you want to load (provided the correct configuration). But, if you host `df` as a `property` then people will be tempted to start implementing additional methods (such as `modify_df` above) as part of `UnifiedIO` subclasses and this means that they will need to use distinct subclasses for each dataset they want to load. Hence, we end up with something like this: 

```python
"""Add module docstring...."""
import os

from demo.src import RESOURCES, ENVIRONMENT, environment
from dynamicio import UnifiedIO
from dynamicio.config import IOConfig
from dynamicio.core import SCHEMA_FROM_FILE


class Bar(UnifiedIO):
    """UnifiedIO subclass for V6 data."""

    schema = SCHEMA_FROM_FILE

    def modify_bar_df(self):
        self.df = self.df.reset_index(drop=True)

    def some_other_modification_for_bar(self):
        ...

class Foo(UnifiedIO):
    """UnifiedIO subclass for V6 data."""

    schema = SCHEMA_FROM_FILE

    def modify_foo_df(self):
        self.df = self.df.reset_index(drop=True)

    def some_other_modification_for_foo(self):
        ...

input_config = IOConfig(
    path_to_source_yaml=(os.path.join(RESOURCES, "definitions/input.yaml")),
    env_identifier=ENVIRONMENT,
    dynamic_vars=environment,
)

bar = Bar(source_config=input_config.get(source_key="BAR"), apply_schema_validations=True, log_schema_metrics=True)
foo = Foo(source_config=input_config.get(source_key="FOO"), apply_schema_validations=True, log_schema_metrics=True)

# do some other stuff
# ...now process bar_df
bar_df = bar.modify_bar_df
foo_df = foo.modify_foo_df

# write output
# finish
```

Again, nothing wrong with this. But it is a change in the framework's philosophy. It begins to function more like a processing framework as well as an I/O framework. Hell, if we define methods to return self, we can even start chaining operations:
```python
class Bar(UnifiedIO):
    """UnifiedIO subclass for V6 data."""

    schema = SCHEMA_FROM_FILE

    def modify_bar_df(self):
        self.df = self.df.reset_index(drop=True)
        return self

    def some_other_modification_for_bar(self):
        ...
        return self

...        

bar_df = bar.modify_bar_df.some_other_modification_for_foo.df 
```

There is then a discussion about whether we should be using `cached_property` or just a `property`. One is immutable and the other is not. 
> A regular property blocks attribute writes unless a setter is defined. In contrast, a cached_property allows writes.

Then, there is this: 
> Note, this decorator interferes with the operation of [PEP 412](https://peps.python.org/pep-0412/) key-sharing dictionaries. This means that instance dictionaries can take more space than usual.

which is imperative to investigate as part of this PR.   
 
All in all, my personal opinion is that we should discuss this thoroughly and decide on the direction. In short, my feedback is: 
* Let's expose one way of loading data (rather than multiple); 
* Let's keep responsibilities aligned with the framework's objective which is to serve as an I/O wrapper (so instances of `DynamicDataIO` should not host data). 

I really hope you are not discouraged by this review and I am looking forward to hearing your thoughts and seeing what others (@tylerferguson , @VRGhost ) have to say.

Reply:
A few points;
1) I agree that it's better to have one way of loading data- issue is that we can't just remove `.read()` as this library is used everywhere, also `.df` uses `.read()` anyway. 
2) While cached_property is supposed to be immutable, since it's a dataframe, both assigning to it and performing in-place actions somehow works. pandas shennanigans. 🤷 
3) Async_read is not a drop-in replacement for `.read()` anyway. You need to call it asynchronously so... ? 
4) Just because you have one property to keep the one dataframe your io object is supposed to load does not mean there would be more data-related functions hosted in said class. 

Happy to discuss, also I don't understand hte PEP412 issue, if you could explain when we discuss please? At the end of the day the outcome of this is just supposed to be a slight syntactic change when using dynamicio.

Reply:
This is actually a huge change that goes against how dynamicio works, fundamentally. A `DynamicDataIO` instance **never** has a `DataFrame` as state. This makes it clear that dynamicio only does reading and writing without keeping references to `DataFrame` objects.

I don't see any benefits bar convenience, and a whole load of drawbacks: unexpected mutability, increased memory use (because the GC can't collect the `DataFrame` as long as you have an active instance of `DynamicDataIO`) and, last but not least, a huge dent in separation of concerns.

Reply:
> I agree that it's better to have one way of loading data- issue is that we can't just remove .read() as this library is used everywhere, also .df uses .read() anyway.

So, since we agree, let's not introduce another way unless it is imperative we do so. 

> While cached_property is supposed to be immutable, since it's a dataframe, both assigning to it and performing in-place actions somehow works. pandas shennanigans. 🤷

That's not true. [It's in the docs](https://docs.python.org/3/library/functools.html#functools.cached_property): 
> The mechanics of [cached_property()](https://docs.python.org/3/library/functools.html#functools.cached_property) are somewhat different from [property()](https://docs.python.org/3/library/functions.html#property). A regular property blocks attribute writes unless a setter is defined. In contrast, a cached_property allows writes. 

Try it with something other than a dataframe. There is no magic/shennanigans  going on here. 

> Async_read is not a drop-in replacement for .read() anyway. You need to call it asynchronously so... ?

My intention with this comment was to emphasize that we already have two ways of reading things. Async reads/writes were introduced as they bring significant value in use cases where one wants to batch- load/write data. Bringing in a third way, which violates the existing paradigms makes things complicated. I am catching myself thinking "If we accept that we want to host dataframes as states (which I don't think we do), then we should also be doing so asynchronously too.". However,  the `df` cached_property your pitching here only works for sychronous reads; what about asynchrounous reads with state? This is just unnecessarily complicated. 

> Just because you have one property to keep the one dataframe your io object is supposed to load does not mean there would be more data-related functions hosted in said class.

Indeed. But it is about what and how people may start using the `DynamicDataIO` instances if it does. Right now, this is not possible. Your alteration makes it possible. With this, we run the risk of blending business logic with I/O logic.

Reply:
Ah true, I confused property and cached_property. Anyway, guess I'll close this convenience.